encoding/ccf/decode_typedef.go

```diff
@@ -0,0 +1,477 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+)
+
+// decodeTypeDefs decodes composite/interface type definitions as
+// language=CDDL
+// composite-typedef = [
+//
+//	+(
+//	  struct-type
+//	  / resource-type
+//	  / contract-type
+//	  / event-type
+//	  / enum-type
+//	  / struct-interface-type
+//	  / resource-interface-type
+//	  / contract-interface-type
+//	  )]
+func (d *Decoder) decodeTypeDefs() (*cadenceTypeByCCFTypeID, error) {
+	// Decode number of type definitions.
+	count, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	if count == 0 {
+		return nil, errors.New("found 0 type definition in composite-typedef (expected at least 1 type definition)")
+	}
+
+	types := newCadenceTypeByCCFTypeID()
+
+	// NOTE: composite fields are not decoded while composite types are decoded
+	// because field type might reference composite type that hasn't decoded yet.
+	rawFields := make(map[ccfTypeID][]byte, count)
+
+	// cadenceTypeIDs is used to check if cadence type IDs are unique in type definitions.
+	cadenceTypeIDs := make(map[cadenceTypeID]struct{}, count)
+
+	// previousCadenceID is used to check if type definitions are sorted by cadence type IDs.
+	var previousCadenceID cadenceTypeID
+
+	for i := uint64(0); i < count; i++ {
+		ccfID, cadenceID, err := d.decodeTypeDef(types, rawFields)
+		if err != nil {
+			return nil, err
+		}
+
+		// "Valid CCF Encoding Requirements" in CCF specs:
+		//
+		//   "composite-type.cadence-type-id MUST be unique in
+		//    ccf-typedef-message or ccf-typedef-and-value-message."
+		if _, ok := cadenceTypeIDs[cadenceID]; ok {
+			return nil, fmt.Errorf("found duplicate Cadence type ID %s in composite-typedef", cadenceID)
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
+		//
+		//   "composite-type.id in ccf-typedef-and-value-message MUST
+		//    be identical to its zero-based index in composite-typedef."
+		if !ccfID.Equal(newCCFTypeIDFromUint64(i)) {
+			return nil, fmt.Errorf(
+				"CCF type ID %d doesn't match composite-typedef index %d in composite-typedef",
+				ccfID,
+				i,
+			)
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
+		//
+		//   "Type definitions MUST be sorted by cadence-type-id in composite-typedef."
+		if !stringsAreSortedBytewise(string(previousCadenceID), string(cadenceID)) {
+			return nil, fmt.Errorf(
+				"Cadence type ID (%s, %s) isn't sorted in composite-typedef",
+				string(previousCadenceID),
+				string(cadenceID),
+			)
+		}
+
+		cadenceTypeIDs[cadenceID] = struct{}{}
+		previousCadenceID = cadenceID
+	}
+
+	// Decode fields after all high-level type definitions are resolved.
+	for id, raw := range rawFields { //nolint:maprange
```

Comment:
Good point! Done in commit f6f08eed16c991f320a671801762f8f79ba30b29.

- replaced map with slice to avoid future problems when metering is added to the function.
- renamed `rawFields` to `rawFieldsOfTypes`
- renamed `raw` to `rawFields`